### PR TITLE
Signal fixes

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

File: glib/Idle.cs [hunk 1]
```diff
@@ -44,44 +50,84 @@ protected override bool Invoke ()
...
-		
+
 		private Idle ()
 		{
 		}
-		
-		[DllImport("libglib-2.0-0.dll", CallingConvention=CallingConvention.Cdecl)]
-		static extern uint g_idle_add (SourceProxy.GSourceFuncInternal d, IntPtr data);
+
+		const int G_PRIORITY_DEFAULT_IDLE = 200;
```
Comment (Therzok) [inline glib/Idle.cs]: Should I add a glue method to invoke on static cctor so we get the value of this define?

Comment (alanmcgovern) [inline glib/Idle.cs]: As a rule of thumb - yes

File: glib/SignalClosure.cs [hunk 2]
```diff
@@ -64,7 +64,7 @@ class SignalClosure : IDisposable
 
 			static Dictionary<IntPtr, SignalClosure> closures = new Dictionary<IntPtr, SignalClosure> (IntPtrEqualityComparer.Instance);
 
-			public SignalClosure (Signal sig, System.Type args_type)
+			public SignalClosure (Signal sig)
```
Comment (alanmcgovern) [inline glib/SignalClosure.cs]: 👍

File: glib/Object.cs [hunk 3]
```diff
@@ -459,7 +459,7 @@ protected void CreateNativeObject (IntPtr [] native_names, GLib.Value [] vals, i
 		[UnmanagedFunctionPointer (CallingConvention.Cdecl)]
 		delegate void NotifyDelegate (IntPtr handle, IntPtr pspec, IntPtr gch);
 
-		void NotifyCallback (IntPtr handle, IntPtr pspec, IntPtr gch)
+		static void NotifyCallback (IntPtr handle, IntPtr pspec, IntPtr gch)
```
Comment (alanmcgovern) [inline glib/Object.cs]: 👍

File: glib/Idle.cs [hunk 4]
```diff
@@ -44,44 +50,84 @@ protected override bool Invoke ()
...
+			IdleProxy proxy = (IdleProxy)gch.Target;
+
+			lock (idle_handlers) {
+				if (proxy.needsAdd)
+					proxy.needsAdd = false;
+				else
+					idle_handlers.Remove (proxy.ID);
+			}
+
+			gch.Free ();
```
Comment (alanmcgovern) [inline glib/Idle.cs]: My concern here is that the `IdleProxy` object still has a reference to the GCHandle and we run the risk of accidentally double-freeing the handle now. Could we make `SourceProxy` IDisposable, then change this code to call `Dispose` on it and null out the GCHandle field as soon as it's freed?

Comment (alanmcgovern) [inline glib/Idle.cs]: Alternatively can you remove the GCHandle field from the object? If we're using a DestroyNotify then the gchandle we pass as `data` will be implicitly destroyed when that function is invoked! If that's the case we can simply move the GCHandle creation code to the place where we pass it to native and set the destroy function. That'll keep things super simple and easy to reason about lifecycles!

File: glib/Idle.cs [hunk 5]
```diff
@@ -36,8 +36,8 @@ public class Idle
 		internal class IdleProxy : SourceProxy
 		{
 			internal readonly IdleHandler real_handler;
-			internal bool needsAdd = true;
 			internal uint ID;
+			internal bool needsAdd = true;
```
Comment (alanmcgovern) [inline glib/Idle.cs]: Let's squash this commit in before we merge!

File: glib/Idle.cs (hunk 4 above)
Comment (Therzok) [inline glib/Idle.cs]: Done the latter.

**Pull request merged**